src/couch_peruser/src/couch_peruser.erl

```diff
@@ -157,54 +250,98 @@ remove_user(User, Prop, {Modified, SecProps}) ->
+               fun (Prop, SAcc) -> TransformFun(User, Prop, SAcc) end,
+               {false, SecProps},
+               [<<"admins">>, <<"members">>]) of
+        {false, _} ->
+            ok;
+        {true, SecProps1} ->
+            ok = fabric:set_security(UserDb, {SecProps1}, [?ADMIN_CTX])
+        end
     end.
 
+-spec user_db_name(User :: binary()) -> binary().
 user_db_name(User) ->
     HexUser = list_to_binary(
         [string:to_lower(integer_to_list(X, 16)) || <<X>> <= User]),
     <<?USERDB_PREFIX,HexUser/binary>>.
 
+-spec exit_changes(State :: #state{}) -> ok.
+exit_changes(State) ->
+    lists:foreach(fun (ChangesState) ->
+        demonitor(State#changes_state.changes_ref, [flush]),
+        unlink(ChangesState#changes_state.changes_pid),
+        exit(ChangesState#changes_state.changes_pid, kill)
+    end, State#state.states).
 
-%% gen_server callbacks
+-spec is_stable() -> true | false.
+is_stable() ->
+    gen_server:call(?MODULE, is_stable).
+
+-spec subscribe_for_changes() -> ok.
+subscribe_for_changes() ->
+    config:subscribe_for_changes([
+        {"couch_httpd_auth", "authentication_db"},
+        "couch_peruser"
+    ]).
+
+% Mem3 cluster callbacks
+
+% TODO: find out what type Server is
+-spec cluster_unstable(Server :: any()) -> any().
+cluster_unstable(Server) ->
+    gen_server:cast(Server, cluster_unstable),
+    Server.
 
+% TODO: find out what type Server is
+-spec cluster_stable(Server :: any()) -> any().
+cluster_stable(Server) ->
+    gen_server:cast(Server, cluster_stable),
+    Server.
+
+%% gen_server callbacks
+-spec init(Options :: list()) -> {ok, #state{}}.
 init([]) ->
     ok = subscribe_for_changes(),
-    {ok, init()}.
+    {ok, init_state()}.
 
+handle_call(is_stable, _From, #state{cluster_stable = IsStable} = State) ->
+    {reply, IsStable, State};
 handle_call(_Msg, _From, State) ->
     {reply, error, State}.
 
 
-handle_cast(update_config, ClusterState) when ClusterState#clusterState.states =/= undefined ->
-    lists:foreach(fun (State) ->
-        demonitor(State#state.changes_ref, [flush]),
-        exit(State#state.changes_pid, kill)
-    end, ClusterState#clusterState.states),
-
-    {noreply, init()};
+handle_cast(update_config, State) when State#state.states =/= undefined ->
+    exit_changes(State),
+    {noreply, init_state()};
 handle_cast(update_config, _) ->
-    {noreply, init()};
+    {noreply, init_state()};
 handle_cast(stop, State) ->
     {stop, normal, State};
+handle_cast(cluster_unstable, State) when State#state.states =/= undefined ->
+    exit_changes(State),
+    {noreply, init_state()};
+handle_cast(cluster_unstable, _) ->
+    {noreply, init_state()};
+handle_cast(cluster_stable, State) ->
+    {noreply, start_listening(State)};
 handle_cast(_Msg, State) ->
     {noreply, State}.
 
-handle_info({'DOWN', Ref, _, _, _Reason}, #state{changes_ref=Ref} = State) ->
+handle_info({'DOWN', Ref, _, _, _Reason}, #changes_state{changes_ref=Ref} = State) ->
```

Comment:
This would be #state because it is the state of the gen_server.

I am not sure about the intent of this clause in general. This is clause would handle a monitor 'DOWN' from a changes feed started with `spawn_opt` with `monitor` enabled that process is also linked, so if it dies unexpectedly this gen_server will also restart. I guess, this monitor would catch the `normal` exits, though a continuous change feed with infinity timeout shouldn't exit?. 

The correct fix in that case we have to check that the Ref is one of the refs in State#state.states then stop, normal. However since we're stopping anyway, maybe it isn't terrible to not even match on Ref and instead just always {stop, normal, State} if any {'DOWN', _Ref, _, _, _Reason} message arrives knowing it has to be one of the change feed processes. Then the only fix would be to replace `#changes_state{changes_ref=Ref} = State` with `State`